core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt

```diff
@@ -0,0 +1,319 @@
+import android.content.Context
+import android.media.AudioManager
+import android.media.AudioManager.OnAudioFocusChangeListener
+import android.os.Build
+import android.os.Build.VERSION_CODES
+import android.speech.tts.TextToSpeech
+import android.speech.tts.TextToSpeech.OnInitListener
+import android.speech.tts.UtteranceProgressListener
+import android.util.Log
+import android.webkit.JavascriptInterface
+import android.webkit.WebView
+import android.widget.Toast
+import org.kiwix.kiwixmobile.core.CoreApp.Companion.instance
+import org.kiwix.kiwixmobile.core.R
+import org.kiwix.kiwixmobile.core.extensions.toast
+import org.kiwix.kiwixmobile.core.reader.ZimReaderContainer
+import org.kiwix.kiwixmobile.core.utils.LanguageUtils.Companion.iSO3ToLocale
+import org.kiwix.kiwixmobile.core.utils.TAG_KIWIX
+import java.util.ArrayList
+import java.util.HashMap
+import java.util.Locale
+import java.util.concurrent.atomic.AtomicInteger
+
+/**
+ * Constructor.
+ *
+ * @param context the context to create TextToSpeech with
+ * @param onInitSucceedListener listener that receives event when initialization of TTS is done
+ * (and does not receive if it failed)
+ * @param onSpeakingListener listener that receives an event when speaking just started or
+ */
+class KiwixTextToSpeech internal constructor(
+  context: Context,
+  onInitSucceedListener: OnInitSucceedListener,
+  onSpeakingListener: OnSpeakingListener,
+  onAudioFocusChangeListener: OnAudioFocusChangeListener,
+  private val zimReaderContainer: ZimReaderContainer
+) {
+  private val focusLock: Any
+  private val context: Context
+  private val onSpeakingListener: OnSpeakingListener
+  private val am: AudioManager
+  private val onAudioFocusChangeListener: OnAudioFocusChangeListener
+  @JvmField var currentTTSTask: TTSTask? = null
+  private var tts: TextToSpeech? = null
+
+  /**
+   * Returns whether the TTS is initialized.
+   *
+   * @return `true` if TTS is initialized; `false` otherwise
+   */
+  private var isInitialized = false
+
+  private fun initTTS(onInitSucceedListener: OnInitSucceedListener) {
+    tts = TextToSpeech(instance, OnInitListener { status: Int ->
+      if (status == TextToSpeech.SUCCESS) {
+        Log.d(TAG_KIWIX, "TextToSpeech was initialized successfully.")
+        this.isInitialized = true
+        onInitSucceedListener.onInitSucceed()
+      } else {
+        Log.e(TAG_KIWIX, "Initialization of TextToSpeech Failed!")
+        Toast.makeText(context, R.string.texttospeech_initialization_failed, Toast.LENGTH_SHORT)
+          .show()
+      }
+    })
+  }
+
+  /**
+   * Reads the currently selected text in the WebView.
+   */
+  fun readSelection(webView: WebView) {
+    webView.loadUrl("javascript:tts.speakAloud(window.getSelection().toString());", null)
+  }
+
+  /**
+   * Starts speaking the WebView content aloud (or stops it if TTS is speaking now).
+   */
+  fun readAloud(webView: WebView) {
+    if (currentTTSTask != null && currentTTSTask!!.paused) {
```

Comment:
```  fun readAloud(webView: WebView) {
    if (currentTTSTask != null) {
      currentTTSTask?.paused.let {
        onSpeakingListener.onSpeakingEnded()
        currentTTSTask = null
      }
 ``` 
@rashiq  Should I approach this?

Reply:
you can remove the  `if (currentTTSTask != null) {` check because the `?` operator already takes care of the null check

Reply:
@rashiq How should we check `currentTTSTask?.paused.`

Reply:
not sure I understand -  just this will do
```
 currentTTSTask?.paused.let {
        onSpeakingListener.onSpeakingEnded()
        currentTTSTask = null
```

Reply:
```
  fun readAloud(webView: WebView) {
    val isPaused = currentTTSTask?.paused
    val isSpeaking = tts?.isSpeaking
    if (isPaused == true) {
      onSpeakingListener.onSpeakingEnded()
      currentTTSTask = null
    } else if (isSpeaking == true) {
```
@rashiq How about this?

Reply:
you don't need `== true`

Reply:
but it's optional boolean